File: src/leetcode/java/ConvertSortedListToBinarySearchTree_109.java
package leetcode.java.medium;

public class ConvertSortedListToBinarySearchTree_109 {
  public TreeNode sortedListToBST(ListNode head) {
    if (head == null) return null;
    if (head.next == null) return new TreeNode(head.val);
    ListNode mid = head;
    ListNode fast = head;
    ListNode beforeMid = null;
    while (fast != null && fast.next != null) {
      fast = fast.next.next;
      beforeMid = mid;
      mid = mid.next;
    }
    if (beforeMid != null) beforeMid.next = null;
    TreeNode node = new TreeNode(mid.val);
    node.left = sortedListToBST(head);
    node.right = sortedListToBST(mid.next);
    return node;
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }

  class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
  }
}

File: src/leetcode/python/ConvertSortedListToBinarySearchTree_109.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class ConvertSortedListToBinarySearchTree_109(object):
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        if not head:
            return None
        if not head.next:
            return TreeNode(head.val)
        mid = fast = head
        before_mid = None
        while head and head.next:
            head = head.next.next
            before_mid = mid
            mid = mid.next
        if before_mid:
            before_mid.next = None
        node = TreeNode(mid.val)
        node.left = TreeNode(head)
        node.right = TreeNode(mid)
        return node
